### ISSUE TITLE
Does .backward() contain grads for both sequences?

### ISSUE THREAD

**Issue by janvainer** [Maghoumi/pytorch-softdtw-cuda#5]
Hi, first of all: Awesome repo! :) Thanks.
To the point: I would like to train an audio-phoneme aligner. I have two encoders.

1. Spectrogram encoder - returns as many embeddings as there are feature frames
2. Phoneme encoder - returns as many embeddings as there are phonemes

I want to align these two sequences with DTW and optimize both the SpecEncoder and PhonemeEncoder.
Is this possible out of the box? I looked in the code but I am not sure if the loss returns grads w.r.t to both input sequences.

**Comment by Maghoumi** [comment]
Hi @janvainer, thanks for your interest!

> I looked in the code but I am not sure if the loss returns grads w.r.t to both input sequences.

Yes, the code can compute the gradients w.r.t both inputs. Let me explain.

Take a look at [_SoftDTWCUDA](https://github.com/Maghoumi/pytorch-softdtw-cuda/blob/master/soft_dtw_cuda.py#L114). This class takes the cost matrix `D` as the input to `forward()`, and computes the sDTW dissimilarity using that input. This class can also compute the gradient of the sDTW function, w.r.t `D` and returns it in `backward()`.

Let's say `D = dist_func(x, y)`. By calling sDTW on D, we are effectively computing the composite function `loss = sDTW( dist_func (x, y))`. Using the chain rule in derivatives, one can compute the derivative of `loss` w.r.t to both its inputs. This explains what you observed in your code snippet above. 

In this code, we're relying on PyTorch's pipeline to apply the chain rule correctly, so overall the computation should be correct.

> Also do you think that this loss could be used to solve forced alignment tasks such as the one I described above?

I'm a bit rusty when it comes to NLP, but if I understood your description correctly, what you intend to do should be possible. 

Hope this helps!

**Comment by janvainer** [comment]
Thanks! helped a lot :)
